### PR TITLE
Early cloud-init

### PR DESCRIPTION
**Pull request by joshwget**
Moves cloud-init save functionality to be run while RancherOS is still running from memory and before System Docker is started. This is done by the following changes.

- The `cloud-init-save` command was refactored so that it could be called by the `init` package
- `cloud-init.sh` was rewritten in Go so it could be called by the `init` package
- The `cloudinit` image was removed since `cloud-init.sh` was the only reason it was needed.
- The `network-pre`, `cloud-init-pre`, and `cloud-init` services were removed. It was common for services to depend on `cloud-init`, so a shim was put in place to move the dependency to `cloud-init-execute`.
- A "cloud-init" service set is introduced. This needs to run only when there's a datasource that requires networking.

For some platforms (notably AWS) we add configuration to disk to define the datasource type. Since cloud-init is now run from memory, we don't have access to this anymore. To get around this RancherOS will now read _only_ the datasource value from disk before running cloud-init. Some refactoring to the `config` package was necessary for this.

Some of the code in `init` is a bit messy due to the odd functional style with `initFunc`. Ideally we can refactor away from this style and then the code quality will be better here.

#1175 
#1275

### CODE REVIEW

File: cmd/cloudinitsave/cloudinitsave.go [hunk 1]
```diff
@@ -46,73 +48,41 @@ const (
...
+	if err := os.MkdirAll(configDevMountPoint, 644); err != nil {
 		return err
 	}
 
-	if err = util.WriteFileAtomic(rancherConfig.MetaDataFile, metaDataBytes, 400); err != nil {
-		return err
+	configDev := util.ResolveDevice(configDev)
+
+	if configDev == "" {
+		return mount.Mount("config-2", configDevMountPoint, "9p", "trans=virtio,version=9p2000.L")
```
Comment (SvenDowideit) [inline cmd/cloudinitsave/cloudinitsave.go]: mmm, you have `configDevMountPoint`, but then use a literal string here?

File: cmd/cloudinitsave/cloudinitsave.go [hunk 2]
```diff
@@ -146,9 +116,67 @@ func saveCloudConfig() error {
...
+		"ec2":          true,
+		"file":         false,
+		"url":          true,
+		"cmdline":      true,
+		"configdrive":  false,
+		"digitalocean": true,
+		"gce":          true,
+		"packet":       true,
+	}[parts[0]]
+	return ok && requiresNetwork
```
Comment (SvenDowideit) [inline cmd/cloudinitsave/cloudinitsave.go]: I feel weird about this being hardcoded in code. What if I'm using `file` but actually need networking?

File: tests/cloud_init_test.go [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+func (s *QemuSuite) TestReadDatasourcesFromDisk(c *C) {
+	s.RunQemu(c)
+
+	s.CheckCall(c, `
+sudo tee /var/lib/rancher/conf/cloud-config.d/datasources.yml << EOF
+rancher:
+  cloud_init:
+    datasources:
+    - url:https://gist.githubusercontent.com/joshwget/e1c49f8b1ddeeba01bc9d0a3be01ed60/raw/9168b380fde182d53acea487d49b680648a0ca5b/gistfile1.txt
```
Comment (SvenDowideit) [inline tests/cloud_init_test.go]: mmm, can we move these into the repo too?

**Comment by SvenDowideit** [comment]
Where's the docs changes :)

File: cmd/cloudinitsave/cloudinitsave.go [hunk 4]
```diff
@@ -46,73 +48,41 @@ const (
...
-
-	dss := getDatasources(cfg)
-	if len(dss) == 0 {
-		return nil, nil
-	}
-
-	ds := selectDatasource(dss)
-	return ds, nil
+func UnmountConfigDrive() error {
+	return syscall.Unmount("/media/config-2", 0)
```
Comment (joshwget) [inline cmd/cloudinitsave/cloudinitsave.go]: This is a reminder that we really should be running a linter in CI :smile:

File: tests/cloud_init_test.go (hunk 3 above)
Comment (joshwget) [inline tests/cloud_init_test.go]: Definitely would be a good change. I'm not sure what would be the best way to go about this though - maybe we can discuss this independently of this PR.

File: cmd/cloudinitsave/cloudinitsave.go (hunk 4 above)
Comment (SvenDowideit) [inline cmd/cloudinitsave/cloudinitsave.go]: especially as i always forget to go fmt.

File: cmd/cloudinitsave/cloudinitsave.go (hunk 1 above)
Comment (joshwget) [inline cmd/cloudinitsave/cloudinitsave.go]: Is this comment saying the same thing as your first comment? I'm not sure what you mean.

**Comment by joshwget** [comment]
Had to update the docs in various places for this. Ended up refactoring the layout a bit and adding a "boot process" section.

File: cmd/cloudinitsave/cloudinitsave.go (hunk 1 above)
Comment (SvenDowideit) [inline cmd/cloudinitsave/cloudinitsave.go]: `"config-2"` seems like another constant.

File: cmd/cloudinitsave/cloudinitsave.go [hunk 5]
```diff
@@ -46,73 +48,42 @@ const (
 	datasourceInterval    = 100 * time.Millisecond
 	datasourceMaxInterval = 30 * time.Second
 	datasourceTimeout     = 5 * time.Minute
+	configDev             = "LABEL=config-2"
+	configDevName         = "config-2"
+	configDevMountPoint   = "/media/config-2"
```
Comment (SvenDowideit) [inline cmd/cloudinitsave/cloudinitsave.go]: so I'd make the derivative 2 look like `configDev= "LABEL="+configDevName` - no big stress tho

File: cmd/cloudinitsave/cloudinitsave.go (hunk 2 above)
Comment (SvenDowideit) [inline cmd/cloudinitsave/cloudinitsave.go]: could be anything - the point is that we're making something that lets other people do what they need - we don't know what that is beforehand..

File: docs/os/boot-process/cloud-init/index.md [hunk 6]
```diff
@@ -0,0 +1,22 @@
+---
+title: Cloud-init
+layout: os-default
+
+---
+
+## Cloud-init
+
+Userdata and metadata can be fetched from a cloud provider, VM runtime, or management service during the RancherOS boot process. Since v0.8.0, this process occurs while RancherOS is still running from memory and before System Docker starts. It is configured by the `rancher.cloud_init.datasources` configuration parameter. For cloud-provider specific images, such as AWS and GCE, the datasource is pre-configured.
```
Comment (SvenDowideit) [inline docs/os/boot-process/cloud-init/index.md]: need to tell the user what "Userdata" and "metadata" are, how they're different, and how the might be used.

**Comment by SvenDowideit** [comment]
basically LGTM - I'm mostly interested in telling the user a little more info in the docs.

File: cmd/cloudinitsave/cloudinitsave.go (hunk 2 above)
Comment (joshwget) [inline cmd/cloudinitsave/cloudinitsave.go]: I don't see this being an issue. If there's a specific use case brought up by users later we can definitely address it then.

**Pull request merged**